Jetbrains update script fixes

AnatolyPopov opened the pull request:
###### Description of changes
Fixing issues caused by this [PR](https://github.com/NixOS/nixpkgs/pull/160115). More specifically:
1. Proper sha256 validation without downloading all the IDEs
2. Proper support of update-channels allowing to upgrade to EAP versions easily.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AnatolyPopov committed:
jetbrains: update script fixes

AnatolyPopov committed:
maintainers: adding AnatolyPopov

AnatolyPopov commented:
@steinybot Could you take a look please since you are the original author of the script?

VergeDX commented:
Ping @steinybot

AnatolyPopov commented on pkgs/applications/editors/jetbrains/versions.json:
```diff
@@ -74,7 +74,7 @@
     "rider": {
       "update-channel": "Rider RELEASE",
-      "url-template": "https://download.jetbrains.com/rider/JetBrains.Rider-{version}.tar.gz",
+      "url-template": "https://download.jetbrains.com/rider/JetBrains.Rider-{versionMajorMinor}-{version}.Checked.tar.gz",
```
Seems to be.

ElXreno replied:
At this point, the links have reverted to the previous template ([example](https://download-cdn.jetbrains.com/rider/JetBrains.Rider-2022.1.tar.gz)). I think this change should be reversed.

AnatolyPopov replied:
Yes, mentioned that today. Fixed.

AnatolyPopov commented on pkgs/applications/editors/jetbrains/update.py:
```diff
@@ -63,18 +62,21 @@ def update_product(name, product):
-            if "sha256" not in product or product.get("version") != version:
-                logging.info("Found a newer version %s.", version)
-                product["version"] = version
+            product["version"] = build["@version"]
```
Seems this line is not needed at all. If `version` requires update the we will get into the `if` statement and update it, otherwise it should not be changed.

AnatolyPopov replied:
Fixed it.

AnatolyPopov commented:
@steinybot I'm just changing the update channel for the package I want to EAP, for example
`"update-channel": "IntelliJ IDEA EAP"`
And after that I'm using update script as before but it updates the version to EAP. 

I'll take a look into your PR a bit later. I was thibnking about something similar(e.g. better support for EAP) but then decided that most of the people anyways will use release versions and support of different update channels is enough.

AnatolyPopov commented:
And current version of script from this branch has almost exactly the same features as old Perl script.

VergeDX commented:
But it seems we all cannot read that Perl code...
i think refactor to Python is a good choice for future changes : )

AnatolyPopov commented:
I was just trying to say that it works as before, nothing really new in my changes :) As for future changes - I really hope to get JetBrains Toolbox working again at some point and all these things will not be needed(at least IMO), so that I don't really care how it's written as soon as it's working as before :)

VergeDX commented:
Maybe you need rebase instead of merge branch from NixOS:master...?

AnatolyPopov commented:
> Maybe you need rebase instead of merge branch from NixOS:master...?

Seems done that accidentally. Removed the merge and rebased on top of master.

ElXreno commented:
@steinybot you can check [this](https://github.com/NixOS/nixpkgs/pull/169398) PR, the update is generated with the updated script.

AnatolyPopov commented:
@steinybot Sure, tested it with both RELEASE and EAP channels since I'm constantly using both.
I still did not manage to take a detailed look at your PR but so far I can see that you are trying to improve these changes further. My suggestion is rather to make it working initially and after that maybe apply your changes.

AnatolyPopov commented:
Changed the script a little bit. The build number needs to be saved to be able to compare EAP builds, so switched the entire version comparison to build numbers and made the version to be more human-readable as was suggested.

AnatolyPopov commented:
Can anybody merge this? I'm not exactly sure who has the permissions for that so maybe you @siraben since you merged the original PR?

VergeDX commented:
Pint @steinybot
